streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -50,65 +57,135 @@
-    private final static String PARAMETER_HINT = "\nYou can increase the producer configs `delivery.timeout.ms` and/or " +
-        "`retries` to avoid this error. Note that `retries` is set to infinite by default.";
-    private final static String TIMEOUT_HINT_TEMPLATE = "%nTimeout exception caught when sending record to topic %s. " +
-            "This might happen if the producer cannot send data to the Kafka cluster and thus, " +
-            "its internal buffer fills up. " +
-            "This can also happen if the broker is slow to respond, if the network connection to " +
-            "the broker was interrupted, or if similar circumstances arise. " +
-            "You can increase producer parameter `max.block.ms` to increase this timeout.";
 
+    // used when eosEnabled is true only
+    private boolean transactionInFlight = false;
+    private Producer<byte[], byte[]> producer;
     private volatile KafkaException sendException;
 
-    public RecordCollectorImpl(final String streamTaskId,
+    public RecordCollectorImpl(final TaskId taskId,
+                               final StreamsConfig config,
                                final LogContext logContext,
-                               final ProductionExceptionHandler productionExceptionHandler,
-                               final Sensor droppedRecordsSensor) {
+                               final StreamsMetricsImpl streamsMetrics,
+                               final Consumer<byte[], byte[]> consumer,
+                               final StreamThread.ProducerSupplier producerSupplier) {
+        this.taskId = taskId;
+        this.consumer = consumer;
         this.offsets = new HashMap<>();
-        this.logPrefix = String.format("task [%s] ", streamTaskId);
         this.log = logContext.logger(getClass());
-        this.productionExceptionHandler = productionExceptionHandler;
-        this.droppedRecordsSensor = droppedRecordsSensor;
+
+        this.applicationId = config.getString(StreamsConfig.APPLICATION_ID_CONFIG);
+        this.productionExceptionHandler = config.defaultProductionExceptionHandler();
+        this.eosEnabled = StreamsConfig.EXACTLY_ONCE.equals(config.getString(StreamsConfig.PROCESSING_GUARANTEE_CONFIG));
+
+        final String threadId = Thread.currentThread().getName();
+        this.droppedRecordsSensor = TaskMetrics.droppedRecordsSensorOrSkippedRecordsSensor(threadId, taskId.toString(), streamsMetrics);
+
+        producer = producerSupplier.get(taskId);
+
+        // TODO: this should be moved to task when it transits to running from created / restoring
+        // then even if there's a long time between the initialization and the first txn that is also fine.
+        initialize();
     }
 
+    /**
+     * @throws StreamsException fatal error that should cause the thread to die
+     */
     @Override
-    public void init(final Producer<byte[], byte[]> producer) {
-        this.producer = producer;
+    public void initialize() {
```

Comment:
req: I'm inclined to remove this unnecessary interface unless for other compelling reason.

Reply:
See my TODO above: my intent is that `initialize` would be called whenever a task transits state, this is because in the past we have a known issue that we call `initTxn` first, and then because of the restoration takes very long time the transaction id actually gets expired even before we call the first `beginTxn`.

I'm leaving it to another PR (supposedly by @vvcephei  :) to do maintain this fix as we did now, which is that when we finished restoring and about to start processing, we initialize the thread's record collector.

Reply:
Sgtm